File: python/day5.py
from functools import total_ordering

from utils import read_lines, test


@total_ordering
class Spine:
    def __init__(self):
        self.left = None
        self.mid = None
        self.right = None
        self.next = None

    @classmethod
    def from_list(cls, list):
        spine = cls()
        for i in list:
            spine.insert(i)
        return spine

    def insert(self, num):
        if not self.mid:
            self.mid = num
        elif num < self.mid and not self.left:
            self.left = num
        elif num > self.mid and not self.right:
            self.right = num
        else:
            if not self.next:
                self.next = Spine()
            self.next.insert(num)

    def __quality(self):
        if not self.next:
            return str(self.mid)
        return str(self.mid) + self.next.__quality()

    def quality(self):
        return int(self.__quality())

    def __score(self):
        score = str(self.mid)
        if self.left:
            score = str(self.left) + score
        if self.right:
            score += str(self.right)
        return int(score)

    def __lt__(self, other):
        if not isinstance(other, Spine):
            return NotImplemented
        if self.quality() == other.quality():
            a = self
            b = other
            while a and b:
                if a.__score() == b.__score():
                    a = a.next
                    b = b.next
                    continue
                return a.__score() < b.__score()
        return self.quality() < other.quality()

    def __eq__(self, other):
        if not isinstance(other, Spine):
            return NotImplemented
        if self.quality() != other.quality():
            return False
        a = self
        b = other
        while a and b:
            if a.__score() != b.__score():
                return False
            a = a.next
            b = b.next
        return True


def get_item(filename):
    return read_lines(filename)


def part1():
    data = get_item(test(5, 1))
    _, nums = data[0].split(":")
    nums = list(map(int, nums.split(",")))
    spine = Spine.from_list(nums)

    return spine.quality()


def part2():
    data = get_item(test(5, 2))
    swords = []
    for line in data:
        _, nums = line.split(":")
        swords.append(
            Spine.from_list(list(map(int, nums.split(",")))).quality()
        )

    best_sword = max(swords)
    worst_sword = min(swords)

    return best_sword - worst_sword


def part3():
    data = get_item(test(5, 3))
    swords = {}
    for line in data:
        i, nums = line.split(":")
        swords[int(i)] = Spine.from_list(list(map(int, nums.split(","))))
    swords = dict(
        sorted(
            swords.items(), key=lambda item: (item[1], item[0]), reverse=True
        )
    )

    return sum([p * i for p, i in enumerate(swords.keys(), start=1)])


def day5():
    print("Day 5:")
    print("\tPart 1:", part1())
    print("\tPart 2:", part2())
    print("\tPart 3:", part3())


if __name__ == "__main__":
    day5()

File: rust/src/day_5.rs
use crate::utils::{Solution, read_lines, test};
use std::path::PathBuf;

#[derive(Debug, Clone, Copy)]
struct Segment {
    mid: u32,
    left: Option<u32>,
    right: Option<u32>,
}

impl Segment {
    #[inline]
    fn new(mid: u32) -> Self {
        Self {
            mid,
            left: None,
            right: None,
        }
    }

    #[inline]
    fn is_full(&self) -> bool {
        self.left.is_some() && self.right.is_some()
    }

    fn try_insert(&mut self, n: u32) -> bool {
        if self.is_full() {
            return false;
        }
        if self.left.is_none() && n < self.mid {
            self.left = Some(n);
            return true;
        }
        if self.right.is_none() && n > self.mid {
            self.right = Some(n);
            return true;
        }
        false
    }

    fn score(&self) -> u64 {
        format!(
            "{}{}{}",
            self.left.map_or(String::new(), |n| n.to_string()),
            self.mid,
            self.right.map_or(String::new(), |n| n.to_string())
        )
        .parse()
        .unwrap()
    }
}

impl PartialEq for Segment {
    fn eq(&self, other: &Self) -> bool {
        self.score() == other.score()
    }
}

impl Eq for Segment {}

impl PartialOrd for Segment {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Segment {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.score().cmp(&other.score())
    }
}

#[derive(Debug, Default)]
struct Spine {
    segments: Vec<Segment>,
}

impl PartialEq for Spine {
    fn eq(&self, other: &Self) -> bool {
        self.quality() == other.quality() && self.segments == other.segments
    }
}

impl Eq for Spine {}

impl PartialOrd for Spine {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Spine {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        if self.quality() == other.quality() {
            return self.segments.cmp(&other.segments);
        }
        self.quality().cmp(&other.quality())
    }
}

impl Spine {
    #[inline]
    fn with_capacity(cap: usize) -> Self {
        Self {
            segments: Vec::with_capacity(cap),
        }
    }

    fn insert(&mut self, n: u32) {
        for segment in &mut self.segments {
            if segment.try_insert(n) {
                return;
            }
        }
        self.segments.push(Segment::new(n));
    }

    fn from_iter<I: IntoIterator<Item = u32>>(iter: I) -> Self {
        let it = iter.into_iter();
        let (lower, _) = it.size_hint();
        let mut spine = Spine::with_capacity(lower);
        for n in it {
            spine.insert(n);
        }
        spine
    }

    fn quality(&self) -> u64 {
        let mut out = String::new();
        for seg in &self.segments {
            out.push_str(seg.mid.to_string().as_str());
        }
        out.parse().unwrap()
    }
}

pub struct Day5;

impl Day5 {
    fn get_test(file: PathBuf) -> Vec<String> {
        read_lines(file).unwrap()
    }
}

impl Solution for Day5 {
    fn part1() -> String {
        let data = Self::get_test(test(5, 1));
        let (_, nums) = data[0].split_once(":").unwrap();
        let spine = Spine::from_iter(nums.split(",").map(|s| s.parse().unwrap()));
        spine.quality().to_string()
    }

    fn part2() -> String {
        let data = Self::get_test(test(5, 2));
        let mut worst_sword = u64::MAX;
        let mut best_sword = 0u64;
        for line in data {
            let (_, nums) = line.split_once(":").unwrap();
            let spine = Spine::from_iter(nums.split(",").map(|s| s.parse().unwrap()));
            worst_sword = worst_sword.min(spine.quality());
            best_sword = best_sword.max(spine.quality());
        }
        (best_sword - worst_sword).to_string()
    }

    fn part3() -> String {
        let data = Self::get_test(test(5, 3));
        let mut swords: Vec<(Spine, u32)> = Vec::new();
        for line in data {
            let (id, nums) = line.split_once(":").unwrap();
            let spine = Spine::from_iter(nums.split(",").map(|s| s.parse().unwrap()));
            swords.push((spine, id.parse().unwrap()));
        }
        swords.sort();
        swords.reverse();
        swords
            .iter()
            .enumerate()
            .map(|(i, v)| (i + 1) as u64 * v.1 as u64)
            .sum::<u64>()
            .to_string()
    }
}
